### ISSUE TITLE
Integer Sampling & Sampler Reduction Modes

### ISSUE THREAD

**Issue by kainino0x** [gpuweb/gpuweb#1267]
It came up while producing [this spreadsheet](https://docs.google.com/spreadsheets/d/12u6g6-dUGxuwgaNVRDdZVZ0u2zb9kotqEO3_ZA-1930/edit?usp=sharing) that it's unclear whether integer formats are ever filterable. @kvark surmised that they might be, in particular with [Sampler Reduction Modes](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VK_EXT_sampler_filter_minmax.html) which allow samplers to reduces samples by `min` or `max` instead of by weighted average.

On Vulkan, at least, `VK_EXT_sampler_filter_minmax` has [surprisingly widespread support](http://vulkan.gpuinfo.org/listdevicescoverage.php?extension=VK_EXT_sampler_filter_minmax) - semi-recent AMD, NVIDIA, Intel, and Qualcomm (some Adreno 500+) all support it.

**Comment by kvark** [comment]
I think the answer here is that "min"/"max" filter modes are not considered "filtering" (just like "nearest" mode isn't filtering).

**Comment by kainino0x** [comment]
In Vulkan at least, they are considered "filtering". However for the purposes of our format capabilities, perhaps not. I wasn't able to easily figure out whether min/max samplers are allowed for integer textures though.

> This extension provides a new sampler parameter which allows applications to produce a filtered texel value by computing a component-wise minimum (MIN) or maximum (MAX) of the texels that would normally be averaged.

**Comment by kvark** [comment]
In Vulkan min/max aren't even filtering modes, they are reduction modes. Here is the VU statement about the filters:

> If a VkSampler created with magFilter or minFilter equal to VK_FILTER_LINEAR and compareEnable equal to VK_FALSE is used to sample a VkImageView as a result of this command, then the image view’s format features must contain VK_FORMAT_FEATURE_SAMPLED_IMAGE_FILTER_LINEAR_BIT

For min/max support there is a separate bit called `VK_FORMAT_FEATURE_SAMPLED_IMAGE_FILTER_MINMAX_BIT`
Vulkan requires the following formats to support it (if they support sampling at all):
>
    VK_FORMAT_R8_UNORM
    VK_FORMAT_R8_SNORM
    VK_FORMAT_R16_UNORM
    VK_FORMAT_R16_SNORM
    VK_FORMAT_R16_SFLOAT
    VK_FORMAT_R32_SFLOAT
    VK_FORMAT_D16_UNORM
    VK_FORMAT_X8_D24_UNORM_PACK32
    VK_FORMAT_D32_SFLOAT
    VK_FORMAT_D16_UNORM_S8_UINT
    VK_FORMAT_D24_UNORM_S8_UINT
    VK_FORMAT_D32_SFLOAT_S8_UINT

So it looks like min/max *is* not a part of "linear filtering"  restriction (which makes sense, it shouldn't be), but whether or not integer formats support min/max is up to the driver to decide and report.

TL;DR: no, we shouldn't plan ahead to support "filtering" on integer formats. The current API should be fine.

**Comment by kainino0x** [comment]
> In Vulkan min/max aren't even filtering modes, they are reduction modes. Here is the VU statement about the filters:

Right, but as it says here filtering must be enabled in order to use them. So they aren't filtering modes, but they're modes that (only) affect filtering.

I missed that MINMAX_BIT. Looks good.

**Comment by kainino0x** [comment]
fwiw, I also found that minmax can't be used with texture swizzling by default.

> If filterMinmaxImageComponentMapping is VK_FALSE the component mapping of the image view used with min/max filtering must have been created with the r component set to the identity swizzle. Only the r component of the sampled image value is defined and the other component values are undefined. If filterMinmaxImageComponentMapping is VK_TRUE this restriction does not apply and image component mapping works as normal.

**Comment by kainino0x** [comment]
or comparison sampling:
> If the format is a depth/stencil format, this bit only specifies that the depth aspect (not the stencil aspect) of an image of this format supports min/max filtering, and that min/max filtering of the depth aspect is supported when depth compare is disabled in the sampler.

**Comment by kvark** [comment]
> fwiw, I also found that minmax can't be used with texture swizzling by default.

yes, we don't have swizzling in WebGPU (yet)

> or comparison sampling

yes, #1283 validates this:
```
         - if |descriptor|.{{GPUSamplerDescriptor/reduction}} is not {{GPUReductionMode/"weighted-average"}},
            then all of the following conditions are satisfied:
            - |device|.{{device/[[features]]}} [=list/contains=] {{GPUFeatureName/"minmax-sampling"}}.
            - |descriptor|.{{GPUSamplerDescriptor/compare}} is `null`.
```

**Comment by kainino0x** [comment]
Unity has graciously provided us with some feedback on this topic:

> > do you use comparison samplers?
> 
> yes we use them for shadows.  we always use the Greater comparison function, we don't get that from the users.
> 
> > do you use min/max filtering?
> 
> as far as I know, we don't use that filtering mode.

Asking for some clarification on whether they assume comparison-filtering results are "good enough".

**Comment by kainino0x** [comment]
More info:
> as far as i know, we only use comparison filters for dx12 shadows, not vulkan or opengl. and we dont do any string checks or runtime tests.

I don't know what the guarantees are of d3d12's comparison filtering so we should look to see if it's stricter than vulkan.
